extensions/kubernetes-client/deployment-internal/src/main/java/io/quarkus/kubernetes/client/deployment/DevServicesKubernetesProcessor.java

```diff
@@ -0,0 +1,415 @@
+        KubernetesDevServicesBuildTimeConfig devServicesConfig = cfg.devservices;
+        return new KubernetesDevServiceCfg(devServicesConfig);
+    }
+
+    private static final class KubernetesDevServiceCfg {
+
+        public boolean devServicesEnabled;
+        public Flavor flavor;
+        public Optional<String> apiVersion;
+        public boolean overrideKubeconfig;
+        public boolean shared;
+        public String serviceName;
+
+        public KubernetesDevServiceCfg(KubernetesDevServicesBuildTimeConfig config) {
+            this.devServicesEnabled = config.enabled;
+            this.serviceName = config.serviceName;
+            this.apiVersion = config.apiVersion;
+            this.overrideKubeconfig = config.overrideKubeconfig;
+            this.flavor = config.flavor;
+            this.shared = config.shared;
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(devServicesEnabled, flavor, apiVersion, overrideKubeconfig, shared, serviceName);
+        }
+
+        @Override
+        public boolean equals(Object obj) {
+            if (this == obj)
+                return true;
+            if (!(obj instanceof KubernetesDevServiceCfg))
+                return false;
+            KubernetesDevServiceCfg other = (KubernetesDevServiceCfg) obj;
+            return devServicesEnabled == other.devServicesEnabled && flavor == other.flavor
+                    && Objects.equals(apiVersion, other.apiVersion) && overrideKubeconfig == other.overrideKubeconfig
+                    && shared == other.shared && Objects.equals(serviceName, other.serviceName);
+        }
+    }
+
+    private class RunningContainer implements ContainerState {
+        private static final String KIND_KUBECONFIG = "/etc/kubernetes/admin.conf";
+        private static final String K3S_KUBECONFIG = "/etc/rancher/k3s/k3s.yaml";
+        private static final String APISERVER = "apiserver";
+        private static final String PKI_BASEDIR = "/etc/kubernetes/pki";
+        private static final String API_SERVER_CA = PKI_BASEDIR + "/ca.crt";
+        private static final String API_SERVER_CERT = PKI_BASEDIR + "/apiserver.crt";
+        private static final String API_SERVER_KEY = PKI_BASEDIR + "/apiserver.key";
+
+        private final DockerClient dockerClient;
+
+        private final InspectContainerResponse containerInfo;
+
+        private final ContainerAddress containerAddress;
+
+        public RunningContainer(DockerClient dockerClient, ContainerAddress containerAddress) {
+            this.dockerClient = dockerClient;
+            this.containerAddress = containerAddress;
+            this.containerInfo = dockerClient.inspectContainerCmd(getContainerId()).exec();
+        }
+
+        public Map<String, String> getKubeconfig() {
+            var image = getContainerInfo().getConfig().getImage();
+            if (image.contains("rancher/k3s")) {
```

Comment:
shouldn't we use the flavor instead of the image to do this logic?

Reply:
Unfortunately, when attaching to a running container we can just describe the container and do some reverse logic to guess what dev services config was used to actually build it. I think of image name or a label applied at creation and choose the former.

Reply:
In theory one could bring their own image, so generally a label is preferable.